libs/amount/amount.cpp

```diff
@@ -0,0 +1,144 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <utility>
+
+#include "amount/amount.hpp"
+
+namespace amount {
+
+  using namespace boost::multiprecision;
+
+  Amount::Amount() : value_(0), precision_(0) {}
+
+  Amount::Amount(uint256_t value)
+      : value_(value), precision_(0) {}
+
+  Amount::Amount(uint256_t amount, uint8_t precision)
+      : value_(amount), precision_(precision) {}
+
+  Amount::Amount(const Amount &am)
+      : value_(am.value_), precision_(am.precision_) {}
+
+  Amount &Amount::operator=(const Amount &other) {
+    // check for self-assignment
+    if (&other == this) return *this;
+    std::copy(&other.value_, &other.value_ + sizeof(decltype(value_)), &value_);
+    std::copy(&other.precision_, &other.precision_ + sizeof(precision_),
+              &precision_);
+    return *this;
+  }
+
+  Amount::Amount(Amount &&am) : value_(0), precision_(0) {
+    value_ = am.value_;
+    precision_ = am.precision_;
+
+    am.value_ = 0;
+    am.precision_ = 0;
+  }
+
+  Amount &Amount::operator=(Amount &&other) {
+    std::swap(value_, other.value_);
+    std::swap(precision_, other.precision_);
+    return *this;
+  }
+
+  Amount Amount::percentage(uint256_t percents) const {
+    uint256_t new_val = value_ * percents / 100;
+    return {new_val, precision_};
+  }
+
+  Amount Amount::percentage(const Amount &am) const {
+    uint256_t new_value = value_ * am.value_;
+    new_value /= uint256_t(std::pow(10, precision_ + am.precision_));
+    return {new_value, precision_};
+  }
+
+  Amount Amount::operator+(const Amount &other) const {
+    if (precision_ != other.precision_) {
+      throw std::invalid_argument("precisions are not the same");
+    }
+    auto new_val = value_ + other.value_;
+    return {new_val, precision_};
+  }
+
+  Amount &Amount::operator+=(const Amount &other) {
+    if (precision_ != other.precision_) {
+      throw std::invalid_argument("precisions are not the same");
+    }
+    value_ += other.value_;
+    return *this;
+  }
+
+  Amount Amount::operator-(const Amount &other) const {
+    if (precision_ != other.precision_) {
+      throw std::invalid_argument("precisions are not the same");
+    }
+    auto new_val = value_ - other.value_;
+    return {new_val, precision_};
+  }
+
+  Amount &Amount::operator-=(const Amount &other) {
+    if (precision_ != other.precision_) {
+      throw std::invalid_argument("precisions are not the same");
+    }
+    value_ -= other.value_;
+    return *this;
+  }
+
+  int Amount::compareTo(const Amount &other) const {
+    if (precision_ == other.precision_) {
+      return (value_ < other.value_) ? -1 : (value_ > other.value_) ? 1 : 0;
+    }
+    // when different precisions transform to have the same scale
+    auto max_precision = std::max(precision_, other.precision_);
+    auto val1 = value_ * uint256_t(std::pow(10, max_precision - precision_));
```

Comment:
integer overflow here, it's better to get the smallest precision

Reply:
IF we get the smallest precision, we would need to do `precision - min_precision`, which would give the same value, so it does not matter to take min or max

Reply:
I mean division instead of mul